File: gdp-docker/gdp/lang/python/gdp/GDP_MD.py
#!/usr/bin/env python

# ----- BEGIN LICENSE BLOCK -----
#	GDP: Global Data Plane
#	From the Ubiquitous Swarm Lab, 490 Cory Hall, U.C. Berkeley.
#
#	Copyright (c) 2015-2019, Regents of the University of California.
#	All rights reserved.
#
#	Permission is hereby granted, without written agreement and without
#	license or royalty fees, to use, copy, modify, and distribute this
#	software and its documentation for any purpose, provided that the above
#	copyright notice and the following two paragraphs appear in all copies
#	of this software.
#
#	IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
#	SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST
#	PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
#	EVEN IF REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#	REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
#	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
#	FOR A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION,
#	IF ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO
#	OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS,
#	OR MODIFICATIONS.
# ----- END LICENSE BLOCK -----

from MISC import *

# metadata id
gdp_md_id_t = c_uint32

class GDP_MD:

    """
    GCL Metadata equivalent -- for internal use only.
    The C equivalent exposed to python users is of a dictionary.

    """

    class gdp_md_t(Structure):
        pass


    def __init__(self, **kwargs):
        """ Constructor: calls the C-function to allocate memory. """

        # We need to be a bit creative here. The python object
        # can be created by allocating new memory for a
        # corresponding gdp_md_t C structure, or it could refer
        # to a gdp_md_t structure created by other means (such
        # as when querying for metadata of a GIN). In the former
        # case, it is our responsibility to clean things up when
        # we no longer need them. However, such automated cleanup
        # may cause problems for the latter case.

        if len(kwargs)==0:
            __func = gdp.gdp_md_new
            __func.argtypes = [c_int]
            __func.restype = POINTER(self.gdp_md_t)

            # Why 0? we can always keep adding entries by calling
            # gdp_md_add (sure it's a bit inefficient, but
            # how often do we need the absolute best performance
            # for gdp_md_t creation?).
            self.ptr = __func(c_int(0))
            self.did_i_create_it = True

        else:
            # The situation when python object was created by
            # passing in a pointer (i.e. memory managed by
            # someone else)
            if "ptr" in kwargs:
                self.ptr = kwargs["ptr"]
                self.did_i_create_it = False
            else:
                raise Exception


    def __del__(self):
        """ Destructor: Frees the C structure """

        if self.did_i_create_it:
            __func = gdp.gdp_md_free
            __func.argtypes = [POINTER(self.gdp_md_t)]

            __func(self.ptr)


    def dump(self, fh, detail, indent):
        """
        Print the gdp_md C memory location contents to a file handle fh.
        fh could be sys.stdout, or any other open file handle.
        Note: This just calls the corresponding C library function which
        handles the printing, details, and indentation.
        """

        # need to convert this file handle to a C FILE*
        __fh = PyFile_AsFile(fh)
        __func = gdp.gdp_md_dump
        __func.argtypes = [POINTER(self.gdp_md_t), FILE_P, c_int, c_int]
        # ignore the return value

        __func(self.ptr, __fh, detail, indent)


    def add(self, md_id, data):
        """ Add a new entry to the metadata set """

        __func = gdp.gdp_md_add
        __func.argtypes = [POINTER(self.gdp_md_t), gdp_md_id_t,
                                c_size_t, c_void_p]
        __func.restype = EP_STAT

        size = c_size_t(len(data))
        tmp_buf = create_string_buffer(data, len(data))

        estat = __func(self.ptr, gdp_md_id_t(md_id),
                                size, byref(tmp_buf))
        check_EP_STAT(estat)


    def get(self, index):
        """ Get a new entry from the metadata set by index """

        __func = gdp.gdp_md_get
        __func.argtypes = [POINTER(self.gdp_md_t), c_int,
                                POINTER(gdp_md_id_t),
                                POINTER(c_size_t), POINTER(c_void_p)]
        __func.restype = EP_STAT

        md_id = gdp_md_id_t()
        dlen = c_size_t()
        data_ptr = c_void_p()
        _index = c_int(index)

        estat = __func(self.ptr, _index, byref(md_id),
                                byref(dlen), byref(data_ptr))
        check_EP_STAT(estat)

        data = string_at(data_ptr, dlen.value)

        return (md_id.value, data)


    def find(self, md_id):
        """ Get a new entry from the metadata set by id """

        __func = gdp.gdp_md_find
        __func.argtypes = [POINTER(self.gdp_md_t), gdp_md_id_t,
                                POINTER(c_size_t), POINTER(c_void_p)]
        __func.restype = EP_STAT

        dlen = c_size_t()
        data_ptr = c_void_p()

        estat = __func(self.ptr, gdp_md_id_t(md_id),
                                byref(dlen), byref(data_ptr))
        check_EP_STAT(estat)

        data = string_at(data_ptr, dlen.value)

        return data

File: combined-docker/gdp/lang/java/org/terraswarm/gdp/GDP_MD.java
/* Global Data Plane meta data.

   Copyright (c) 2015-2016 The Regents of the University of California.
   All rights reserved.
   Permission is hereby granted, without written agreement and without
   license or royalty fees, to use, copy, modify, and distribute this
   software and its documentation for any purpose, provided that the above
   copyright notice and the following two paragraphs appear in all copies
   of this software.

   IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
   FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
   ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
   THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
   SUCH DAMAGE.

   THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
   PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
   CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
   ENHANCEMENTS, OR MODIFICATIONS.

   PT_COPYRIGHT_VERSION_2
   COPYRIGHTENDKEY

 */

package org.terraswarm.gdp; 

import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.awt.PointerInfo;
import java.lang.Exception;


import com.sun.jna.Native;
import com.sun.jna.Memory;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;
import org.terraswarm.gdp.NativeSize; // Fixed by cxh in makefile.

/**
 * Global Data Plane meta data.
 * 
 * @author Nitesh Mor
 */

public class GDP_MD {

    // See the C-API for an explanation of why these magic numbers were chosen.

    public static final int GDP_MD_XID          = 0x00584944; // XID (external id)
    public static final int GDP_MD_PUBKEY       = 0x00505542; // PUB (public key, deprecated)
    public static final int GDP_MD_OWNERPUBKEY  = 0x004F504B; // OPK (owner public key)
    public static final int GDP_MD_WRITERPUBKEY = 0x0057504B; // WPK (writer public key)
    public static final int GDP_MD_CTIME        = 0x0043544D; // CTM (creation time)
    public static final int GDP_MD_EXPIRE       = 0x0058544D; // XTM (expiration date/time)
    public static final int GDP_MD_CREATOR      = 0x00434944; // CID (creator id)
    public static final int GDP_MD_SYNTAX       = 0x0053594E; // SYN (data syntax: json, xml, etc.)
    public static final int GDP_MD_LOCATION     = 0x004C4F43; // LOC (location: lat/long)
    public static final int GDP_MD_NONCE        = 0x004E4F4E; // NON (unique nonce)

    
    /**
     * Pass a pointer to an already allocated gdp_md_t C structure.
     * @param d A pointer to an existing gdp_md_t
     */
    public GDP_MD(PointerByReference d) {
        this.gdp_md_ptr = d;
        this.did_i_create_it = false;
        return;
    }
    
    /**
     * If called without an existing pointer, we allocate a new 
     * C gdp_md_t structure.
     */
    public GDP_MD() {
        this.gdp_md_ptr = Gdp21Library.INSTANCE.gdp_md_new(0);
        this.did_i_create_it = true;
    }
 
    /**
     * If called without an existing pointer with non-zero entries,
     * we allocate a new C gdp_md_t structure with appropriate args.
     */
    public GDP_MD(int entries) {
        this.gdp_md_ptr = Gdp21Library.INSTANCE.gdp_md_new(entries);
        this.did_i_create_it = true;
    }
    
    /**
     * In case we allocated memory for the C gdp_md_t ourselves, free it.
     */
    public void finalize() {        
        if (this.did_i_create_it) {
            Gdp21Library.INSTANCE.gdp_md_free(this.gdp_md_ptr);
        }
    }
    
    /**
     * Add a new entry to the metadata
     * @param md_id  ID for the metadata. Internally a 32-bit unsigned integer
     * @param data  The corresponding data. We automatically calculate length
     *              and pass to the corresponding C routine.
     */
    public void add(int md_id, byte[] data) throws GDPException {

        EP_STAT estat;
        
        // FIXME: figure out if we need to have a +1 for len
        NativeSize len = new NativeSize(data.length);

        // Create a C string that is a copy (?) of the Java String.
        Memory memory = new Memory(data.length+1);
        
        // FIXME: not sure about alignment.
        Memory alignedMemory = memory.align(4);
        memory.clear();
        Pointer pointer = alignedMemory.share(0);

        for (int i=0; i<data.length; i++) {
            pointer.setByte(i,data[i]);
        }
                
        estat = Gdp21Library.INSTANCE.gdp_md_add(this.gdp_md_ptr, md_id,
                len, pointer);

        // don't forget error checking
        GDP.check_EP_STAT(estat, "GDP_MD.java: Adding metadata entry (" +
                                        md_id + ") failed.");
    }

    
    /**
     * Get a metadata entry by index. This is not searching by a md_id. For
     * that, look at find.
     * @param index The particular metadata entry we are looking for
     * @return  Returns a Key-Value pair containing the requested index. 
     *          A hashMap seems to be the best choice, since there's no tuple
     *          type in Java.
     */ 
    public HashMap<Integer, byte[]> get(int index) throws GDPException {

        EP_STAT estat;

        // Pointer to hold the md_id
        IntBuffer md_id_ptr = IntBuffer.allocate(1);
        
        // to hold the length of the buffer
        NativeSize len = new NativeSize();
        
        // Pointer to the buffer to hold the value
        PointerByReference p = new PointerByReference();
        
        estat = Gdp21Library.INSTANCE.gdp_md_get(this.gdp_md_ptr, 
                index, md_id_ptr, new NativeSizeByReference(len), p);

        GDP.check_EP_STAT(estat, "GDP_MD.java: gdp_md_get(" + index + ") failed.");
        
        // get the md_id
        int md_id = md_id_ptr.get(0);
        
        Pointer pointer = p.getValue();
        byte[] val = pointer.getByteArray(0, len.intValue());
        
        HashMap<Integer, byte[]> ret = new HashMap<Integer, byte[]>();
        
        ret.put(md_id, val);
        
        return ret;
    }
    
    /**
     * Find a particular metadata entry by looking up md_id
     * @param md_id ID we are searching for
     * @return the value associated with <code>md_id</code> 
     */
    public byte[] find(int md_id) throws GDPException {

        EP_STAT estat;

        // to hold the length of the buffer
        NativeSize len = new NativeSize();
        // Pointer to the buffer to hold the value
        PointerByReference p = new PointerByReference();
        
        estat = Gdp21Library.INSTANCE.gdp_md_find(this.gdp_md_ptr, 
                md_id, new NativeSizeByReference(len), p);

        GDP.check_EP_STAT(estat, "GDP_MD.java: gdp_md_find(" + md_id + ") failed.");
        
        Pointer pointer = p.getValue();
        byte[] bytes = pointer.getByteArray(0, len.intValue());
        return bytes;
        
    }

    // Pointer to the C structure
    public PointerByReference gdp_md_ptr = null;
    
    // To keep track of whether we need to call free in destructor
    private boolean did_i_create_it = false;


}
